module/zcommon/zfs_fletcher.c

```diff
@@ -245,11 +214,258 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+static const struct fletcher_4_calls fletcher_4_generic_calls = {
+	.init = fletcher_4_generic_init,
+	.compute = fletcher_4_generic,
+	.compute_byteswap = fletcher_4_generic_byteswap,
+	.name = "generic"
+};
+
+extern struct fletcher_4_calls fletcher_4_avx2_calls;
+
+static const struct fletcher_4_calls *fletcher_4_algos[] = {
+	&fletcher_4_generic_calls,
+#if defined(HAVE_AVX) && defined(HAVE_AVX2)
+	&fletcher_4_avx2_calls,
+#endif
+};
+
+#define	ZFS_ARRAY_SIZE(a) 	(sizeof (a) / sizeof (*a))
+
+static const struct fletcher_4_calls *
+    fletcher_4_chosen = &fletcher_4_generic_calls;
+static const struct fletcher_4_calls *fletcher_4_fastest = NULL;
+
+static enum fletcher_selector {
+	FLETCHER_FASTEST = 0,
+	FLETCHER_ORIGINAL,
+	FLETCHER_CYCLE
+} fletcher_4_impl_seletor = FLETCHER_ORIGINAL;
+
+static const char *selector_names[] = {
+	"fastest",
+	"original",
+#if !defined(_KERNEL)
+	"cycle",
+#endif
+};
+
+static kmutex_t fletcher_4_impl_seletor_lock;
+
+static kstat_t *fletcher_4_kstat;
+
+static kstat_named_t fletcher_4_kstat_data[ZFS_ARRAY_SIZE(fletcher_4_algos)];
```

Comment:
All of these global variables need to be declared at the top of this file.

Tunings, such as `fletcher_4_impl_set`, need to have a  reasonable comment describing exactly what they control and why you might want to change this.  New module options also need to have a similar section added to `man/man5/zfs-module-parameters.5` which is easy for end users to understand.